### PR TITLE
feat: vpn blocking

### PR DESCRIPTION
**Pull request by nikkaroraa**
Fixes ENG-1896 ENG-1949

## /blocked page
Screenshot 2023-11-23 at 5 30 55 PM

### CODE REVIEW

File: pages/_middleware.ts [hunk 1]
```diff
@@ -1,23 +1,51 @@
...
-  const country = req?.geo?.country
-  const response = NextResponse.next()
-  const path = req.url
-
-  if (country && BLOCKED_COUNTRIES.includes(country) && !ALLOWED_URLS.test(path)) {
-    return NextResponse.rewrite('/blocked')
+import { redis } from '../utils/redisClient'
+import { isVPN } from '../utils/vpn'
+
+const ignoredPaths = ['/api', '/favicon.ico', '/static', '/_next', '/blocked']
```
Comment (aleone) [inline pages/_middleware.ts:6]: We dealing with /api in other PR? Can discuss if any qs

File: pages/_middleware.ts [hunk 2]
```diff
@@ -1,23 +1,51 @@
...
+    const redisData = await redis.get(ip)
+    const isIPBlocked = !!redisData
+
+    console.log('ip', ip, isIPBlocked, url.protocol, url.host, '/blocked')
+
+    if (isIPBlocked && url.pathname !== '/blocked') {
+      return NextResponse.redirect(`${url.protocol}//${url.host}/blocked`)
+    }
+
+    const isFromVpn = await isVPN(ip)
```
Comment (aleone) [inline pages/_middleware.ts:35]: should implement the same things from data center PRs, but can be in separate PR if prefer

File: pages/_middleware.ts [hunk 3]
```diff
@@ -1,23 +1,51 @@
...
+    if (isFromVpn && url.pathname !== '/blocked') {
+      await redis.set(ip, 1)
```
Comment (aleone) [inline pages/_middleware.ts:37]: same comments here as in squeethmonorepo PR about resetting this storage and force checking the VPN at some regular cadence, but can be in separate PR of course

**Review by aleone — APPROVED** [review]
generally lgtm, should just do similar things as squeeth monorepo here, but can be in separate PRs

make sure we coordinate merge + if need any whitelisted IPs do in coordinated way

File: pages/_middleware.ts (hunk 2 above)
Comment (nikkaroraa) [reply pages/_middleware.ts]: this commit takes care of it: https://github.com/KMKoushik/squeeth-portal/pull/118/commits/b24ecec64bc61571acd033e09ccaa54444f6f137

File: pages/_middleware.ts (hunk 3 above)
Comment (nikkaroraa) [reply pages/_middleware.ts]: done!

File: pages/_middleware.ts (hunk 1 above)
Comment (nikkaroraa) [reply pages/_middleware.ts]: done in this commit: https://github.com/KMKoushik/squeeth-portal/pull/118/commits/09c56044f8809bd44a501de437af13252193579e
